### PR TITLE
Assignment9+10 @Hieu Tran

### PR DESCRIPTION
**Pull request by DakotaCondos**

### CODE REVIEW

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 1]
```diff
@@ -59,64 +61,88 @@ public void RunTaskAsync_Success()
...
+        try
+        {
+            PingResult result = task.Result;
+        }
+        catch (AggregateException exception)
+        {
+            exception = exception.Flatten();
+            exception.Handle(innerException =>
+            {
+                ExceptionDispatchInfo.Capture(innerException).Throw();
```
Comment (dcox2005) [inline Assignment/Assignment.Tests/PingProcessTests.cs:118]: This is a nice and different way to solve this from how we solved it. I like this more.

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 2]
```diff
@@ -59,64 +61,88 @@ public void RunTaskAsync_Success()
...
     async public Task RunAsync_MultipleHostAddresses_True()
     {
         // Pseudo Code - don't trust it!!!
         string[] hostNames = new string[] { "localhost", "localhost", "localhost", "localhost" };
-        int expectedLineCount = PingOutputLikeExpression.Split(Environment.NewLine).Length*hostNames.Length;
+        int expectedLineCount = PingOutputLikeExpression.Split(Environment.NewLine).Length*hostNames.Length; //logic not correct
         PingResult result = await Sut.RunAsync(hostNames);
+        Console.WriteLine(result.StdOutput);
         int? lineCount = result.StdOutput?.Split(Environment.NewLine).Length;
         Assert.AreEqual(expectedLineCount, lineCount);
```
Comment (dcox2005) [inline Assignment/Assignment.Tests/PingProcessTests.cs:133]: This is how we solved it.  Not sure why you commented that the logic is not correct.

File: Assignment/Assignment/PingProcess.cs [hunk 3]
```diff
@@ -20,49 +15,52 @@ public PingResult Run(string hostNameOrAddress)
...
     }
 
     async public Task<PingResult> RunAsync(
         string hostNameOrAddress, CancellationToken cancellationToken = default)
     {
-        Task task = null!;
-        await task;
-        throw new NotImplementedException();
-    }
+        cancellationToken.ThrowIfCancellationRequested();
```
Comment (mmwoodfo) [inline Assignment/Assignment/PingProcess.cs:32]: You should put this within your Task.Run

File: Assignment/Assignment/PingProcess.cs [hunk 4]
```diff
@@ -20,49 +15,52 @@ public PingResult Run(string hostNameOrAddress)
...
+        Task<PingResult> task = Task.Run(() => Run(hostNameOrAddress), cancellationToken);
 
-    async public Task<PingResult> RunAsync(params string[] hostNameOrAddresses)
+        PingResult result = await task;
+        return result;
```
Comment (mmwoodfo) [inline Assignment/Assignment/PingProcess.cs:36]:
```suggestion
     return await task.WaitAsync(cancellationToken);
```

File: Assignment/Assignment/PingProcess.cs [hunk 5]
```diff
@@ -20,49 +15,52 @@ public PingResult Run(string hostNameOrAddress)
...
-        return new PingResult(total, stringBuilder?.ToString());
+        int total = all.Aggregate(0, (total, item) => total + item.Result.ExitCode);
+        stringBuilder.Append(all.Aggregate("", (resultString, currentString) => resultString.Trim() + currentString.Result.StdOutput));
+        return new PingResult(total, stringBuilder?.ToString().Trim());
     }
 
-    async public Task<PingResult> RunLongRunningAsync(
-        string hostNameOrAddress, CancellationToken cancellationToken = default)
+    public Task<PingResult> RunLongRunningAsync(ProcessStartInfo startInfo,
+        Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token)
```
Comment (mmwoodfo) [inline Assignment/Assignment/PingProcess.cs:54]: You've got some unused paramters here

File: Assignment/Assignment/PingProcess.cs [hunk 6]
```diff
@@ -20,49 +15,52 @@ public PingResult Run(string hostNameOrAddress)
...
     {
-        Task task = null!;
-        await task;
-        throw new NotImplementedException();
+
+        Task<PingResult> task = Task.Factory.StartNew(() => Run(startInfo.Arguments), token,
```
Comment (mmwoodfo) [inline Assignment/Assignment/PingProcess.cs:57]: You should be throwing when cancled.

**Pull request closed without merging**